File: bayer-chatbot/internal/logger/logger.go
package logger

import (
	"encoding/json"
	"os"
	"strings"
	"time"
)

type Level int

const (
	LevelSilent Level = iota
	LevelError
	LevelWarn
	LevelInfo
	LevelDebug
)

type Logger struct {
	level Level
}

func New(level string) *Logger {
	lvl := parseLevel(level)
	return &Logger{level: lvl}
}

func parseLevel(level string) Level {
	s := strings.ToLower(strings.TrimSpace(level))
	switch s {
	case "silent":
		return LevelSilent
	case "error":
		return LevelError
	case "warn":
		return LevelWarn
	case "debug":
		return LevelDebug
	case "info":
		fallthrough
	default:
		return LevelInfo
	}
}

func (l *Logger) Enabled(level Level) bool {
	return l.level >= level && l.level != LevelSilent
}

func (l *Logger) Debug(msg string, fields map[string]interface{}) {
	l.log(LevelDebug, "debug", msg, fields)
}

func (l *Logger) Info(msg string, fields map[string]interface{}) {
	l.log(LevelInfo, "info", msg, fields)
}

func (l *Logger) Warn(msg string, fields map[string]interface{}) {
	l.log(LevelWarn, "warn", msg, fields)
}

func (l *Logger) Error(msg string, fields map[string]interface{}) {
	l.log(LevelError, "error", msg, fields)
}

func (l *Logger) log(level Level, levelName, msg string, fields map[string]interface{}) {
	if !l.Enabled(level) {
		return
	}

	payload := map[string]interface{}{
		"ts":    time.Now().UTC().Format(time.RFC3339Nano),
		"level": levelName,
		"msg":   msg,
	}
	for k, v := range fields {
		payload[k] = v
	}

	b, err := json.Marshal(payload)
	if err != nil {
		return
	}

	// Keep it simple: stdout for info/debug, stderr for warn/error.
	if level == LevelError || level == LevelWarn {
		_, _ = os.Stderr.Write(append(b, '\n'))
		return
	}
	_, _ = os.Stdout.Write(append(b, '\n'))
}

File: bayer-chatbot/legacy-ts/src/logger.ts
export type LogLevel = "silent" | "error" | "warn" | "info" | "debug";

const levelOrder: Record<LogLevel, number> = {
  silent: 0,
  error: 1,
  warn: 2,
  info: 3,
  debug: 4
};

export type LoggerOptions = {
  level: LogLevel;
  pretty?: boolean;
};

export class Logger {
  private readonly level: LogLevel;
  private readonly pretty: boolean;

  constructor(options: LoggerOptions) {
    this.level = options.level;
    this.pretty = options.pretty ?? true;
  }

  child(fields: Record<string, unknown>) {
    return new ChildLogger(this, fields);
  }

  enabled(level: LogLevel) {
    return levelOrder[this.level] >= levelOrder[level];
  }

  log(level: Exclude<LogLevel, "silent">, msg: string, fields?: Record<string, unknown>) {
    if (!this.enabled(level)) return;

    const payload = {
      ts: new Date().toISOString(),
      level,
      msg,
      ...(fields ?? {})
    };

    const line = this.pretty ? JSON.stringify(payload) : String(payload);
    if (level === "error") console.error(line);
    else if (level === "warn") console.warn(line);
    else console.log(line);
  }

  error(msg: string, fields?: Record<string, unknown>) {
    this.log("error", msg, fields);
  }

  warn(msg: string, fields?: Record<string, unknown>) {
    this.log("warn", msg, fields);
  }

  info(msg: string, fields?: Record<string, unknown>) {
    this.log("info", msg, fields);
  }

  debug(msg: string, fields?: Record<string, unknown>) {
    this.log("debug", msg, fields);
  }
}

class ChildLogger {
  private readonly parent: Logger;
  private readonly fields: Record<string, unknown>;

  constructor(parent: Logger, fields: Record<string, unknown>) {
    this.parent = parent;
    this.fields = fields;
  }

  error(msg: string, fields?: Record<string, unknown>) {
    this.parent.error(msg, { ...this.fields, ...(fields ?? {}) });
  }

  warn(msg: string, fields?: Record<string, unknown>) {
    this.parent.warn(msg, { ...this.fields, ...(fields ?? {}) });
  }

  info(msg: string, fields?: Record<string, unknown>) {
    this.parent.info(msg, { ...this.fields, ...(fields ?? {}) });
  }

  debug(msg: string, fields?: Record<string, unknown>) {
    this.parent.debug(msg, { ...this.fields, ...(fields ?? {}) });
  }

  enabled(level: LogLevel) {
    return this.parent.enabled(level);
  }
}
